core/include/gap/core/dense_map.hpp

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2022-present, Trail of Bits, Inc.
+
+#pragma once
+
+#include <gap/core/concepts.hpp>
+#include <utility>
+#include <vector>
+
+namespace gap
+{
+
+    //
+    // Dense map stores map from integers to values in a flat container.
+    //
+    // It keeps empty slots for unmapped keys.
+    // NOTE: The implementation uses default value as a flag to denote missing
+    // value in the map.
+    //
+    template<
+        gap::unsigned_integral key_t,
+        typename value_t,
+        typename container_t = std::vector< value_t > >
```

Comment:
Maybe it would be worth to formalise what `container_t` must support as concept?